### PR TITLE
 PeerSharingServer refactorisation #4782 

### PR DESCRIPTION
**Pull request by bolt12**
Refactor Peer Sharing

Refactors PeerSharing `peerSharingServer` to provide a better interface to consensus to fill via `PeerSharingAPI`

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs [hunk 1]
```diff
@@ -104,11 +115,99 @@ peerSharingClient controlMessageSTM
...
+                     )
+                  => PeerSharingAPI peer s m
                   -> PeerSharingServer peer m
-peerSharingServer computePeersToShare =
+peerSharingServer psAPI =
   PeerSharingServer
-    { recvMsgShareRequest = \amount -> (,) <$> computePeersToShare amount
-                                          <*> return (peerSharingServer computePeersToShare)
+    { recvMsgShareRequest = \amount -> (,) <$> computePeerSharingPeers psAPI amount
+                                          <*> return (peerSharingServer psAPI)
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:128]:
```suggestion
    { recvMsgShareRequest = \amount -> (,) <$> computePeerSharingPeers psAPI amount
                                           <*> return (peerSharingServer psAPI)
```

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs [hunk 2]
```diff
@@ -104,11 +115,99 @@ peerSharingClient controlMessageSTM
...
+--
+data PeerSharingAPI addr s m =
+  PeerSharingAPI { psPublicPeerSelectionStateVar :: StrictTVar m (PublicPeerSelectionState addr)
+                 , psGenVar                      :: StrictTVar m s
+                 , psReSaltAtVar                 :: StrictTVar m Time
+                 , psPolicyPeerShareStickyTime   :: !DiffTime
+                 -- ^ Amount of time between changes to the salt used to pick peers to
+                 -- gossip about.
+                 , psPolicyPeerShareMaxPeers     :: !PeerSharingAmount
+                 -- ^ Maximum number of peers to respond with in a single request
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:145]: Do we want these things to be configurable?

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection/MockEnvironment.hs [hunk 3]
```diff
@@ -574,12 +574,9 @@ mockPeerSelectionPolicy GovernorMockEnvironment {
       policyPeerShareBatchWaitTime     = 3,    -- seconds
       policyPeerShareOverallTimeout    = 10,   -- seconds
       policyPeerShareActivationDelay   = 300,  -- seconds
-      policyPeerShareStickyTime        = 257,  -- seconds
-      policyPeerShareMaxPeers          = 10,
       policyErrorDelay              = 10    -- seconds
```
Comment (coot) [inline ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection/MockEnvironment.hs:577]: Could you align the above line?

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs [hunk 4]
```diff
@@ -104,11 +115,99 @@ peerSharingClient controlMessageSTM
...
+
+--
+-- Utility Function
+--
+
+-- | PeerSharingAPI needed to compute the peers to be shared.
+--
+data PeerSharingAPI addr s m =
+  PeerSharingAPI { psPublicPeerSelectionStateVar :: StrictTVar m (PublicPeerSelectionState addr)
+                 , psGenVar                      :: StrictTVar m s
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:139]: Is there any benefit for `s`? We'll anyway just use `StdGen` (which can be used in both `IO` and `IOSim`), isn't it?

**Review by coot — COMMENTED** [review]
Nice, just a few comments.

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs (hunk 4 above)
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:139]: unfortunately `s` is needed due to some polymorphism in the test suite.

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs (hunk 2 above)
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:145]: That's a good question. I don't know? Maybe?

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:145]: `psPolicyPeerShareStickyTime` could be somewhat useful if made configurable.
A lower value would let a nodes peers be spread around more quickly. Possibly useful if one operate a peer with lots of known peers.

`psPolicyPeerShareMaxPeers` same reason, for a node with lots of know peers

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:145]: I will create an issue to make this configurable.

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:145]: https://github.com/IntersectMBO/ouroboros-network/issues/4801

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs [hunk 5]
```diff
@@ -104,11 +115,99 @@ peerSharingClient controlMessageSTM
...
-                  => (PeerSharingAmount -> m [peer])
+peerSharingServer :: ( MonadSTM m
+                     , MonadMonotonicTime m
+                     , Hashable peer
+                     , RandomGen s
+                     )
+                  => PeerSharingAPI peer s m
                   -> PeerSharingServer peer m
-peerSharingServer computePeersToShare =
+peerSharingServer psAPI =
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:125]: Let's not short cut names.
```suggestion
peerSharingServer peerSharingAPI =
```

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE NamedFieldPuns   #-}
+{-# LANGUAGE RecordWildCards  #-}
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:3]: Let's remove `RecordWildCards`; they hide where arguments are coming from and going to.

Don't we have it in our style guide?

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:3]: That's a good question, from a quick grep we use it on other modules

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSharing.hs]: I know, in a few places. I'd rather remove it than expand it.

**Review by coot — APPROVED** [review]
LGTM

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs [hunk 7]
```diff
@@ -104,11 +114,104 @@ peerSharingClient controlMessageSTM
...
+newPeerSharingAPI rng policyPeerShareStickyTime policyPeerShareMaxPeers = do
+  publicPeerSelectionStateVar <- newTVarIO emptyPublicPeerSelectionState
+  genVar <- newTVarIO rng
+  reSaltAtVar <- newTVarIO (Time 0)
+  return $
+    PeerSharingAPI publicPeerSelectionStateVar
+                   genVar
+                   reSaltAtVar
+                   policyPeerShareStickyTime
+                   policyPeerShareMaxPeers
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:176]: Could you use `NamedFieldPuns` - it makes maintenance easier.

File: ouroboros-network/src/Ouroboros/Network/PeerSharing.hs [hunk 8]
```diff
@@ -104,11 +114,104 @@ peerSharingClient controlMessageSTM
...
+                           )
+                        => PeerSharingAPI ntnAddr s m
+                        -> PeerSharingAmount
+                        -> m [ntnAddr]
+computePeerSharingPeers PeerSharingAPI{ psPublicPeerSelectionStateVar
+                                      , psPolicyPeerShareStickyTime
+                                      , psPolicyPeerShareMaxPeers
+                                      , psReSaltAtVar
+                                      , psGenVar
+                                      } amount = do
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSharing.hs:196]: according to our style guide:
```suggestion
computePeerSharingPeers PeerSharingAPI { psPublicPeerSelectionStateVar,
                                         psPolicyPeerShareStickyTime,
                                         psPolicyPeerShareMaxPeers,
                                         psReSaltAtVar,
                                         psGenVar
                                       } amount = do
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 9]
```diff
@@ -668,8 +663,7 @@ runM Interfaces
     (policyRng,      rng2) = split rng1
     (churnRng,       rng3) = split rng2
     (fuzzRng,        rng4) = split rng3
-    (ntnInbgovRng,   rng5) = split rng4
-    (ntcInbgovRng,   peerSharingRng) = split rng5
+    (ntnInbgovRng, ntcInbgovRng)    = split rng4
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:666]:
```suggestion
    (ntnInbgovRng, ntcInbgovRng) = split rng4
```

**Review by coot — APPROVED** [review]

**Pull request merged**